mapie/calibration.py

```diff
@@ -0,0 +1,502 @@
+from typing import Optional, Union, cast, Tuple, Dict
+
+import warnings
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin, clone, RegressorMixin
+from sklearn.calibration import _SigmoidCalibration
+from sklearn.isotonic import IsotonicRegression
+from sklearn.pipeline import Pipeline
+from sklearn.model_selection import BaseCrossValidator
+from sklearn.utils import check_random_state, check_consistent_length
+from sklearn.utils.multiclass import type_of_target
+from sklearn.utils.validation import (
+    indexable,
+    check_is_fitted,
+    _check_y,
+    _num_samples,
+)
+
+from ._typing import ArrayLike, NDArray
+from .utils import (
+    check_cv,
+    check_null_weight,
+    check_n_features_in,
+    check_calib_set,
+    fit_estimator,
+    check_estimator_classification,
+    check_estimator_fit_predict,
+    check_binary_zero_one,
+)
+
+
+class MapieCalibrator(BaseEstimator, ClassifierMixin):
+
+    fit_attributes = [
+        "estimator",
+        "calibrators",
+    ]
+
+    valid_calibrators = {
+        "sigmoid": _SigmoidCalibration(),
+        "isotonic": IsotonicRegression(out_of_bounds="clip")
+    }
+
+    def __init__(
+        self,
+        estimator: Optional[ClassifierMixin] = None,
+        method: str = "top_label",
+        calibrator: Optional[Union[str, RegressorMixin]] = None,
+        cv: Optional[Union[str, BaseCrossValidator]] = "split",
+    ) -> None:
+        self.estimator = estimator
+        self.method = method
+        self.calibrator = calibrator
+        self.cv = cv
+
+    def _fit_calibrator(
+        self,
+        item: int,
+        calibrator: RegressorMixin,
+        y_calib: ArrayLike,
+        top_class_prob: NDArray,
+        top_class_prob_arg: NDArray,
+        sample_weight: Optional[ArrayLike],
+    ) -> RegressorMixin:
+        """
+        Fitting the calibrator requires that for each class we
+        get the correct values to fit a calibrator for this specific
+        class.
+
+        Parameters
+        ----------
+        item : int
+            The class for which we will fit a calibrator.
+        calibrator : RegressorMixin
+            Calibrator to train.
+        y_calib : NDArray of shape (n_samples,)
+            The dependent values of the calibrator.
```

Comment:
Training labels.
to be consistent with _fit_calibrators

Reply:
done.